### PR TITLE
Allow arbitrary attributes & methods for `ThisEmulator`

### LINKED ISSUE #6253: Support for DBT's `this.identifier`
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

As can be seen in `jinja.py` ([here](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/templaters/jinja.py#L300)), `ThisEmulator` only supports `name`, `schema`, and `database`. But it should also support `identifier`, as documented [here](https://docs.getdbt.com/reference/resource-properties/identifier), as well as `type`, `is_table`, `is_view`, `is_cte`, and `include()`, as documented [here](https://docs.getdbt.com/reference/dbt-classes#creating-relations). 

Basically, `ref()`, `source()`, and `this` all return a Relation object.

image

### Use case

Not get SQLFluff errors when using `this.identifier`.

### Dialect

Jinja/DBT

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by mroy-seedbox**
### Brief summary of the change made
- Allows to use `{{ this.identifier }}` in DBT (but with the Jinja templater).
- Also allows any other future attributes or methods.
- Rename `ThisEmulator` to `RelationEmulator`.
- Also return a RelationEmulator for `ref()` and `source()` as well.
- Fixes #6253

### Are there any other side effects of this change that we should be aware of?
No. Identical as before, but supports more possibilities.

### Sample results
```python
t = ThisEmulator()
print(t)
print(t.database)
print(t.schema)
print(t.name)
print(t.identifier)
print(t.type)
print(t.something_new)
print(t.is_table)
print(t.is_view)
print(t.is_materialized_view)
print(t.is_cte)
print(t.is_dynamic_table)
print(t.is_iceberg_format)
print(t.include())
print(t.include(database=False))
print(t.some_new_method())
```

Prints:
```
this_model
this_database
this_schema
this_model
this_model
this_model
this_model
True
True
True
True
True
True
this_model
this_model
this_model
```

### Alternative Solutions
Initially, I tried this:
```
[sqlfluff:templater:jinja:context]
this = {"identifier": "this_identifier"}
```

But then that broke `{{ this }}`. 🤦‍♂️ So not a solution.

Then I tried this:
```
[sqlfluff:templater:jinja:context]
this.identifier = "this_identifier"
```

Which simply does not work at all. 😞 But... it would be pretty awesome if it did work! It would give us a lot more flexibility to customize things (in order to override only specific parts of objects, like `this.*`, `target.*`, and others).

### CODE REVIEW

**Comment by mroy-seedbox** [comment]
Also changed the return of `ref()` and `source()` to a `RelationEmulator`.

### Sample results
```python
t = ThisEmulator("ref_model")
print(t)
print(t.database)
print(t.schema)
print(t.name)
print(t.identifier)
print(t.type)
print(t.something_new)
print(t.is_table)
print(t.is_view)
print(t.is_materialized_view)
print(t.is_cte)
print(t.is_dynamic_table)
print(t.is_iceberg_format)
print(t.include())
print(t.include(database=False))
print(t.some_new_method())
```

Prints:
```
ref_model
this_database
this_schema
ref_model
ref_model
ref_model
ref_model
True
True
True
True
True
True
ref_model
ref_model
ref_model
```

**Review by alanmcruickshank — COMMENTED** [review]
Neat! I like the approach. Will review properly once I've seen what it does to the test suite, but on a first pass I think this looks like a good approach.

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: I fixed a few things! The next test run should work better. 🤞

**Comment by mroy-seedbox** [comment]
Ah crap.... `typing.Self` was only added in Python 3.11. 🤦‍♂️

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: Fixed!

File: src/sqlfluff/core/templaters/jinja.py [hunk 1]
```diff
@@ -297,21 +299,40 @@ def _generate_dbt_builtins():
...
             """A class which emulates the `this` class from dbt."""
 
-            name = "this_model"
+            identifier = "this_model"
             schema = "this_schema"
             database = "this_database"
 
+            def __init__(
+                self, identifier: str = "this_model"
+            ): -> None  # pragma: no cover TODO?
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/templaters/jinja.py:311]: Typo
```suggestion
            ) -> None:  # pragma: no cover TODO?
```

**Comment by mroy-seedbox** [comment]
Thanks for the quick fix! 🙌 

Not sure why it can't find `typing_extensions`... I added it to [requirements_dev.txt](https://github.com/sqlfluff/sqlfluff/pull/6254/files#diff-c9c796259f3852b51b531b79cbf07820145088d4f108fb006ab44b15f0b15934). 🤷‍♂️

**Comment by mroy-seedbox** [comment]
Ah, duh.... it's not just a dev dependency. 🤦‍♂️

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: Just added `typing-extensions` to `pyproject.toml`, so that should fix the remaining issues!

**Comment by alanmcruickshank** [comment]
> @alanmcruickshank: Just added `typing-extensions` to `pyproject.toml`, so that should fix the remaining issues!

I'm not sure you should need to add `typing-extensions`, I've been wrestling a lot with typing recently so have a suggestion on that front. I've also removed the `# pragma: no cover` comments, because I think we should try and get coverage on those methods.

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: Tests are ready!

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: Just need to run the test coverage on the latest commit, and it should pass! 👌

**Comment by mroy-seedbox** [comment]
Actually, just checking this now, and there's a little improvement that could be made to conform better to how DBT Relations work!

Won't take long!

**Comment by mroy-seedbox** [comment]
The full test suite is still running on my end, but everything is looking good so far!

**Comment by mroy-seedbox** [comment]
@alanmcruickshank: Alright, all good on my end! ✅  Although I can't get the coverage report to complete. It says "No data to combine". 🤷‍♂️ 

Apologies for my laziness & lack of time/effort earlier in this process. 😞

**Review by alanmcruickshank — APPROVED** [review]
Nice. Looks great. Thanks for contributing this 👍

**Pull request merged**